src/ucp/core/ucp_mm.c

```diff
@@ -1196,3 +1294,365 @@ ucs_status_t ucp_mem_reg_md_map_update(ucp_context_h context)
+                                     UCP_MEM_FLAG_IMPORTED, &memh);
+        if (status != UCS_OK) {
+            goto err_rcache_destroy;
+        }
+
+        user_memh->parent = memh;
+    } else {
+        memh   = user_memh;
+        rcache = NULL;
+    }
+
+    memh->reg_id      = reg_id;
+    memh->remote_uuid = remote_uuid;
+    status            = ucp_memh_import_attach(context, memh,
+                                               attach_params_array,
+                                               attach_params_num);
+    if (status != UCS_OK) {
+        goto err_memh_free;
+    }
+
+out:
+    UCP_THREAD_CS_EXIT(&context->mt_lock);
+    return status;
+
+err_memh_free:
+    if (rcache != NULL) {
+        ucs_rcache_region_put_unsafe(rcache, &memh->super);
+    }
+err_rcache_destroy:
+    if ((rcache != NULL) && (rcache != existing_rcache)) { 
+        /* Rcache was allocated here - remove it from hash and destroy */
+        iter = kh_get(ucp_context_imported_mem_rcaches_hash,
+                      context->imported_mem_rcaches, remote_uuid);
+        ucs_assert(iter != kh_end(context->imported_mem_rcaches));
+        kh_del(ucp_context_imported_mem_rcaches_hash,
+               context->imported_mem_rcaches, iter);
+        ucs_rcache_destroy(rcache);
+    }
+    goto out;
+}
+
+static ucs_status_t
+ucp_memh_import(ucp_context_h context, const void *export_mkey_buffer,
+                ucp_mem_h *memh_p)
+{
+    ucs_rcache_t *rcache = NULL;
+    const void *p        = export_mkey_buffer;
+    const void *end;
+    ucp_md_index_t md_index;
+    ucp_mem_h memh, rcache_memh;
+    ucp_md_index_t remote_md_index;
+    ucs_status_t status;
+    ucp_md_map_t remote_md_map, local_md_map;
+    ucp_memh_import_attach_params_t *attach_params_array;
+    unsigned attach_params_num;
+    ucs_memory_type_t mem_type;
+    uint64_t UCS_V_UNUSED flags;
+    uint64_t UCS_V_UNUSED mem_info_parsed_size;
+    uint64_t remote_uuid;
+    const void *tl_mkey_buf;
+    ucs_rcache_region_t *rregion;
+    khiter_t iter;
+    void *address;
+    size_t length;
+    uint64_t memh_info_size;
+    uint64_t reg_id;
+
+    ucs_assert(export_mkey_buffer != NULL);
+
+    /* Common memory handle information */
+    memh_info_size = *ucs_serialize_next(&p, uint16_t);
+    end            = UCS_PTR_BYTE_OFFSET(export_mkey_buffer, memh_info_size);
+
+    flags          = *ucp_memh_serialize_next(&p, uint64_t, end, 0lu);
+    remote_md_map  = *ucp_memh_serialize_next(&p, ucp_md_map_t, end, 0u);
+    mem_type       = *ucp_memh_serialize_next(&p, uint8_t, end,
+                                              UCS_MEMORY_TYPE_HOST);
+
+    /* Exported memory handle stuff */
+    address     = (void*)*ucp_memh_serialize_next(&p, uint64_t, end, 0lu);
+    length      = *ucp_memh_serialize_next(&p, uint64_t, end, 0lu);
+    remote_uuid = *ucp_memh_serialize_next(&p, uint64_t, end, 0lu);
+    reg_id      = *ucp_memh_serialize_next(&p, uint64_t, end, UINT64_MAX);
```

Comment:
how is that possible these values are not packed? why need to use this new serialize macro?